File: gecko/csrc/ops/cema_blelloch_scan.cc
#include "ops/cema_blelloch_scan.h"

namespace gecko{
namespace ops{

std::tuple<torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor>
CEMABlellochScanFwd(
    const torch::Tensor& x,
    const torch::Tensor& p,
    const torch::Tensor& q,
    const torch::Tensor& gamma,
    const c10::optional<torch::Tensor>& bos_mask,
    const c10::optional<torch::Tensor>& h0){
  return CEMABlellochScanCUDAFwd(x, p, q, gamma, bos_mask, h0);
}

std::tuple<torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor>
CEMABlellochScanBwd(
    const torch::Tensor& y_grad,
    const c10::optional<torch::Tensor>& h_last_grad,
    const torch::Tensor& chunk_q_prefix,
    const torch::Tensor& chunk_h_prefix,
    const torch::Tensor& x,
    const torch::Tensor& p,
    const torch::Tensor& q,
    const torch::Tensor& gamma,
    const c10::optional<torch::Tensor>& bos_mask){
  return CEMABlellochScanCUDABwd(y_grad, h_last_grad, chunk_q_prefix, chunk_h_prefix, x, p, q, gamma, bos_mask);
}

void DefineCEMABlellochScanOp(py::module& m) {
    m.def("cema_blelloch_scan_fwd", &CEMABlellochScanFwd, "CEMABlellochScanFwd")
        .def("cema_blelloch_scan_bwd", &CEMABlellochScanBwd, "CEMABlellochScanBwd");
}

}  // namespace ops
}  // namespace gecko

File: gecko/modules/fused_ops/cema_blelloch_scan.py
from typing import Optional, Tuple
import torch
from torch.autograd.function import FunctionCtx

from gecko_extension.ops import (
    cema_blelloch_scan_fwd,
    cema_blelloch_scan_bwd,
    cema_cub_scan_fwd,
    cema_cub_scan_bwd,
)

cema_scan_fwd = cema_cub_scan_fwd
cema_scan_bwd = cema_cub_scan_bwd


class CEMABlellochScanFunc(torch.autograd.Function):

    @staticmethod
    def forward(
        ctx: FunctionCtx,
        x:torch.Tensor,
        hx: Optional[torch.Tensor],
        p: torch.Tensor,
        q: torch.Tensor,
        gamma: torch.Tensor,
        bos_mask: Optional[torch.Tensor],
    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
        y, h, chunk_decay, chunk_gain = cema_scan_fwd(x, p, q, gamma, bos_mask, hx)
        ctx.save_for_backward(x, hx, p, q, gamma, chunk_decay, chunk_gain, bos_mask)
        return y, h

    @staticmethod
    def backward(
        ctx: FunctionCtx,
        y_grad: torch.Tensor,
        h_grad: Optional[torch.Tensor]
    ) -> Tuple[torch.Tensor, Optional[torch.Tensor],
               torch.Tensor, torch.Tensor, torch.Tensor, None]:
        x, hx, p, q, gamma, chunk_decay, chunk_gain, bos_mask = ctx.saved_tensors
        x_grad, p_grad, q_grad, gamma_grad, hx_grad = cema_scan_bwd(
            y_grad, h_grad, chunk_decay, chunk_gain, x, p, q, gamma, bos_mask
        )
        hx_grad = None if hx is None else hx_grad

        return x_grad, hx_grad, p_grad, q_grad, gamma_grad, None


cema_blelloch_scan = CEMABlellochScanFunc.apply
